### PR TITLE
Implement TryFrom<&[T]> for &[T; N]

### PR DESCRIPTION
**Pull request by nvzqz**
There are many cases where a buffer with a static compile-time size is preferred over a slice with a dynamic size. This allows for performing a checked conversion from `&[T]` to `&[T; N]`. This may also lead to compile-time optimizations involving `[T; N]` such as loop unrolling.

This is my first PR to Rust, so I'm not sure if discussion of this change should happen here or does it need its own RFC? I figured these changes would be a subset of #33417.

### CODE REVIEW

**Comment by shepmaster** [comment]
Thanks for the awesome PR, @nvzqz! We'll wrangle up a top-notch reviewer for you in a short amount of time!

**Comment by jethrogb** [comment]
I'd love to make arrays better, but shouldn't we wait for const generics?

**Comment by shepmaster** [comment]
> shouldn't we wait for const generics?

Is there a reason this would *conflict* or not be implementable with const generics? Otherwise it seems to leave us in no worse shape than we are now.

**Comment by jethrogb** [comment]
> Is there a reason this would conflict

Not that I'm aware of. People have previously been cautious about metadata bloat though.

**Comment by nvzqz** [comment]
This change adds 66 new implementations of `TryFrom`, so I understand the concern about metadata bloat. However, I don't want to wait until integer generics to have this change. It seems likely that `TryFrom` will be stabilized way before that.

**Comment by BurntSushi** [comment]
I haven't been following `TryFrom` or const generics, so I'd like to hand this off to someone else. @sfackler?

**Comment by alexcrichton** [comment]
ping @sfackler, mind taking a look at this?

File: src/libcore/array.rs
Comment (sfackler) [inline src/libcore/array.rs]: This should probably implement `Error` in libstd.

**Comment by sfackler** [comment]
I'm on board with these existing, but let's see if everyone is okay with more ad-hoc array stuff.

@rfcbot fcp merge

**Comment by sfackler** [comment]
Oops, forgot to libs tag it

@rfcbot fcp merge

File: src/libcore/array.rs
Comment (nvzqz) [reply src/libcore/array.rs]: Should the `Display` implementation also be in libstd or libcore?

Comment (nvzqz) [reply src/libcore/array.rs]: See e45e8abec2302530d774aab1ca800081b60aa89e

**Comment by bluss** [comment]
`slice.len() >= $N` would still be valid, but I guess "slicing" to an array shorter than the input slice is not compatible with the spirit of the `TryFrom` trait (?).

**Comment by nvzqz** [comment]
I believe that the `TryFrom` conversions should be equal to the length of the slice. You can always do:

```rust
let values = [1, 2, 3, 4];

let three: &[i32; 3] = values[..3].try_into().unwrap();
```

**Comment by sfackler** [comment]
In other similar situations like copy_from_slice and clone_from_slice, we require that the lengths match exactly.

**Comment by scottmcm** [comment]
If `TryFrom` starts allowing `&[T] -> &]T; N]`, shouldn't `From` support `&[T; N] -> &[T]` for symmetry?

**Comment by SimonSapin** [comment]
`&[T; N]` already coerce to `&[T]` implicitly in many places, so unless some specific code could use a `From` impl in a generic context, that impl doesn’t seem very useful. `&[T]` to `&[T; N]` on the other hand is new functionality that previously required unsafe code and macros to achieve.

**Comment by shepmaster** [comment]
ping @aturon for your ticky boxes [here](https://github.com/rust-lang/rust/pull/44764#issuecomment-332930367)!

**Comment by nvzqz** [comment]
Is this just waiting on @aturon? It's been 3 weeks now 😕

**Comment by carols10cents** [comment]
@aturon was out for two weeks, but hopefully should be back now, but he is traveling a bit again this week :) i'll ping him on IRC.

File: src/libcore/array.rs
Comment (bjorn3) [inline src/libcore/array.rs]: Why not use the description method from the `Error` trait?

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: src/libcore/array.rs
Comment (nvzqz) [reply src/libcore/array.rs]: Error is declined in libstd, not libcore

Comment (nvzqz) [reply src/libcore/array.rs]: Defined* (I'm on mobile)

Comment (bjorn3) [reply src/libcore/array.rs]: I see, never mind.

**Comment by alexcrichton** [comment]
Egad sorry for the delay! This can now be approved after we've all signed off --

@bors: r+

**Comment by jethrogb** [comment]
We're missing `impl From<Infallible> for TryFromSliceError`, see #60318

**Pull request merged**
